src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp

```diff
@@ -37,29 +37,50 @@ namespace stan {
       }
 
       // The time derivative of the virial, G = \sum_{d = 1}^{D} q^{d} p_{d}.
-      virtual double dG_dt(Point& z) = 0;
+      virtual double dG_dt(
+        Point& z,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) = 0;
 
       // tau = 0.5 p_{i} p_{j} Lambda^{ij} (q)
-      virtual const Eigen::VectorXd dtau_dq(Point& z) = 0;
+      virtual const Eigen::VectorXd dtau_dq(
+        Point& z,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) = 0;
 
       virtual const Eigen::VectorXd dtau_dp(Point& z) = 0;
 
       // phi = 0.5 * log | Lambda (q) | + V(q)
-      virtual const Eigen::VectorXd dphi_dq(Point& z) = 0;
+      virtual const Eigen::VectorXd dphi_dq(
```

Comment:
If the virtual overhead's too high here, we could probably go with a CRTP implementation for the Hamiltonians.  I only bring this up because this is a pretty inner-loop kind of code.